### PR TITLE
feat(security): add JWT authorization

### PR DESCRIPTION
**Pull request by Edholm**
## Description

Built to work with auth0, but should theoretically work with any provider.

## How has this been tested?

- [x] Unit tests
- [x] Manually

## Screenshots
Not really screenshottable...

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -18,6 +20,8 @@ require (
 require (
 	github.com/agnivade/levenshtein v1.1.0 // indirect
 	github.com/cpuguy83/go-md2man/v2 v2.0.0-20190314233015-f79a8a8ca69d // indirect
+	github.com/dgrijalva/jwt-go v3.2.0+incompatible // indirect
+	github.com/form3tech-oss/jwt-go v3.2.5+incompatible // indirect
```
Comment (ericalofstrom) [inline go.mod:24]: Check: How come you used these jwt libraries? On jwt.io you can see list of libraries that are nice to use: https://jwt.io/

**Comment by Edholm** [comment]
@abjugard @ericalofstrom please have a look when you get the time. #security

File: internal/cinema/cinemaserver.go [hunk 2]
```diff
@@ -63,8 +72,13 @@ func (s *Server) Routes(ctx context.Context) *mux.Router {
 	r := mux.NewRouter()
 
 	// Middleware
+	// TODO: request id?
+	r.Use(middleware.RecoverPanic())
 	r.Use(middleware.AttachAppLogger(logger))
 	r.Use(middleware.ProcessMaintenance(s.cfg))
+	r.Use(middleware.ApplyAuthorization(s.cfg))
+	r.Use(middleware.ApplyAuthentication())
```
Comment (ericalofstrom) [inline internal/cinema/cinemaserver.go:80]: Check: Why is authentication after authorization. Does the order matter?

**Review by ericalofstrom — APPROVED** [review]
Generally looks good, can't really see any obvious issues except that a lot of tests are missing :D

File: go.mod (hunk 1 above)
Comment (Edholm) [reply go.mod:24]: those libs are used indirectly. I think dgrijalva/jwt-go is an old version of golang-jwt/jwt that along with MicahParks/keyfunc.

But yes, I should add tests... Wanna help 😉 ?

File: internal/cinema/cinemaserver.go (hunk 2 above)
Comment (Edholm) [reply internal/cinema/cinemaserver.go:80]: Yes, order matters. They are executed in that order. Authentication requires authorization to provide the token.

Comment (ericalofstrom) [reply internal/cinema/cinemaserver.go:80]: But doesn't that mean that you could get a 403 when you should get a 401?

File: go.mod (hunk 1 above)
Comment (ericalofstrom) [reply go.mod:24]: Yeah I'd like to help if I just knew how to write them :D

Comment (Edholm) [reply go.mod:24]: Easy to learn 😆

File: internal/cinema/cinemaserver.go (hunk 2 above)
Comment (ericalofstrom) [reply internal/cinema/cinemaserver.go:80]: Kind of what authentication means, but as I see it authentication is that you have a JWT, and authorization is that the JWT contains the correct claims, such as scopes. 

1. Validera att anropet är korrekt
2. Validera att access token är korrekt
3. Transformera access token till rättighetsmodell
4. Validera att data i anropet är korrekt
5. Validera behörighet att utföra operationen
6. Validera behörighet till det data som efterfrågas eller påverkas

Step 2 in this list is: 
* Korrekt kryptografisk signatur
* Signerad av rätt IdP (oftast konfigurerad som en URL till vår IdP)
* Utställd med en audience som gäller för vårt API
* Korrekt typ, dvs. en access token och inte någonting annat
* Giltig, t ex med avseende på tid

But authorization for me is when you have for example "movies.create" scope.

Comment (ericalofstrom) [reply internal/cinema/cinemaserver.go:80]: But I trust that you know that it's in the correct order, I have no clue :grin:

Comment (Edholm) [reply internal/cinema/cinemaserver.go:80]: hm I don't think we're actually verifying that we get an access token and not an id_token,

but 1 and 2 is the authorization middleware in this PR, 3 is basically what the authentication middleware is going to do (transform the access token to an id_token).

I will look into checking the type though...

**Review by abjugard — APPROVED** [review]

**Pull request merged**